src/qudi/hardware/fast_adc/spectrum/si_commands/si_commands.py

```diff
@@ -0,0 +1,128 @@
+        self.ts_buf = TsBufferCommands(card)
+        self.cfg = ConfigureCommands(card, log)
+        self.process = ProcessAction(self)
+
+
+class ProcessAction:
+    """
+    This class has actions used in the data process. They modify the commands to be useful in the measurement.
+    """
+    def __init__(self, commands:Commands):
+        self.cmd = commands
+
+        self.wait_time_interval = 0.01
+
+        self.trigger_enabled = False
+        self._seq_size_B = 0
+        self._total_gate = 0
+        self._ts_seq_size_B = 0
+
+    def input_ms(self, ms: MeasurementSettings):
+        self._seq_size_B = ms.seq_size_B
+        self._assign_methods(ms.gated)
+        self._total_gate = ms.total_gate
+        if ms.gated:
+            self._ts_seq_size_B = ms.ts_seq_size_B
+
+    def _assign_methods(self, gated):
+        if gated:
+            self.get_trig_reps = self._get_trig_reps_gated
+            self.get_curr_avail_reps = self._get_curr_avail_reps_gated
+        else:
+            self.get_trig_reps = self._get_trig_reps_ungated
+            self.get_curr_avail_reps = self._get_curr_avail_reps_ungated
+
+
+    def get_avail_user_reps(self):
+        return int(np.floor(self.cmd.data_buf.get_avail_user_len_B() / self._seq_size_B))
+
+    def get_ts_avail_user_reps(self):
+        return int(self.cmd.ts_buf.get_ts_avail_user_len_B() / self._ts_seq_size_B)
+
+    def toggle_trigger(self, trigger_on):
+        if trigger_on == self.trigger_enabled:
+            return
+        else:
+            if trigger_on:
+                self.trigger_enabled = self.cmd.card.enable_trigger()
+            else:
+                self.trigger_enabled = self.cmd.card.disable_trigger()
+
+    def wait_new_trig_reps(self, prev_trig_reps):
+        curr_trig_reps = self.get_trig_reps()
+        while curr_trig_reps == prev_trig_reps:
+            curr_trig_reps = self.get_trig_reps()
+            time.sleep(self.wait_time_interval)
+        return curr_trig_reps
+
+    def _get_trig_reps_ungated(self):
+        return int(self.cmd.data_buf.get_trig_counter())
+
+    def _get_trig_reps_gated(self):
+        return int(self.cmd.data_buf.get_trig_counter() / self._total_gate)
+
+    def _get_curr_avail_reps_ungated(self):
+        return self.cmd.data_buf.get_avail_user_reps()
+
+    def _get_curr_avail_reps_gated(self):
+        curr_data_avail_reps = self.cmd.process.get_avail_user_reps()
+        curr_ts_avail_reps = self.cmd.process.get_ts_avail_user_reps()
+        return min(curr_data_avail_reps, curr_ts_avail_reps)
+
+    def wait_avail_data(self):
+        initial_time = time.time()
+        curr_avail_reps = self.get_curr_avail_reps()
+        while curr_avail_reps == 0:
+            curr_avail_reps = self.get_curr_avail_reps()
+            current_time = time.time() - initial_time
+            if current_time > 10:
+                time_out = True
+                return time_out
+            time.sleep(self.wait_time_interval)
```

Comment:
This might also lock the thread for a significant amount of time, making it unresponsive. If wait_time_interval is too long this might cause problems with stopping the acquisition.

Reply:
I removed this wait function from the measurement loop(si_commander).
Now, instead of waiting in another while loop(wait_new_trig_reps), it returns in the measurement loop.

``` 
def command_process(self):
        """
        Command the main process dependent on the repetitions of unprocessed data.
        """

        trig_reps = self.cmd.process.get_trig_reps()
        processed_reps = self.processor.data.avg.num_rep
        unprocessed_reps = trig_reps - processed_reps
        if self._reps != 0 and processed_reps >= self._reps:
            return
```